SSA/Core/MLIRSyntax/Parser.lean

```diff
@@ -0,0 +1,138 @@
+  | Except.error e => e.toMessageData.toString
+
+unsafe def elabIntoEIO {α : Type} (env : Lean.Environment) (typeName : Lean.Expr) (stx : Lean.Syntax) : EIO ParseError α :=
+  fun s =>
+    let resE : EIO Exception α := elabIntoCore (α := α) typeName stx |>.run' {fileName := "parserHack", fileMap := default} {env := env}
+    match resE s with
+    | .ok a s => .ok a s
+    | .error exception s =>
+      let errMsgIO : IO String := exception.toMessageData.toString
+      match errMsgIO s with
+      | .ok errMsg s => .error s!"failed elaboration {stx}. Error: {errMsg}" s
+      | .error e s => .error s!" Failed elaborating {stx}.\nUnable to pretty-print exception at 'elabIntoEIO':\n{e}." s
+
+def region0Expr := (Expr.app (Expr.const `MLIR.AST.Region []) (Expr.const `Nat.zero []))
+def op0Expr := (Expr.app (Expr.const `MLIR.AST.Op []) (Expr.const `Nat.zero []))
+/--
+  Parse `Lean.Syntax` into `MLIR.AST.Region`.
+
+  This runs the Lean frontend stack and adds it to the TCB.
+  It should be safe (if we trust the lean parser) since we ensure that the types match: see
+  https://leanprover-community.github.io/mathlib4_docs/Std/Util/TermUnsafe.html#Std.TermUnsafe.termUnsafe_
+ -/
+def elabRegion (env : Lean.Environment) (stx : Lean.Syntax) : EIO ParseError (MLIR.AST.Region 0) := do
+  let reg ← unsafe elabIntoEIO (α := MLIR.AST.Region 0) env region0Expr stx
+  return reg
+
+/--
+  Parse `Lean.Syntax` into `MLIR.AST.Op`.
+
+  This runs the Lean frontend stack and adds it to the TCB.
+  It should be safe (if we trust the lean parser) since we ensure that the types match: see
+  https://leanprover-community.github.io/mathlib4_docs/Std/Util/TermUnsafe.html#Std.TermUnsafe.termUnsafe_
+ -/
+def elabOp (env : Lean.Environment) (stx : Lean.Syntax) : EIO ParseError (MLIR.AST.Op 0) := do
+  let reg ← unsafe elabIntoEIO (α := MLIR.AST.Op 0) env op0Expr stx
+  return reg
+
+def mkParseFun (syntaxcat : Name)
+    (ntparser : Syntax → EIO ParseError ParseOutput)
+    (s : String) (env : Environment) : EIO ParseError ParseOutput := do
+  let syn : Except String Syntax := Parser.runParserCategory env syntaxcat s
+  match syn with
+    | .error msg => throw msg
+    | .ok syn => ntparser syn
+
+private def mkNonTerminalParser (syntaxcat : Name)
+    (ntparser : Syntax → EIO ParseError ParseOutput)
+    (env : Environment) (s : String) : EIO ParseError ParseOutput :=
+  let parseFun := mkParseFun syntaxcat ntparser
+  parseFun s env
+
+-- TODO: do we need this copy of the environment, or can I remove it from one of the two?
+def regionParser : @ParseFun (MLIR.AST.Region 0) :=
+  fun env : Lean.Environment => mkNonTerminalParser `mlir_region (elabRegion env) env
+
+private def parseFile (env: Lean.Environment)
+    (parser: @ParseFun ParseOutput)
+    (filepath: System.FilePath) : IO (Option ParseOutput) := do
+  let lines ← IO.FS.lines filepath
+  let fileStr := lines.foldl (λ s₁ s₂ => s₁ ++ "\n" ++ s₂) "" -- TODO: use intercalate.
```

Comment:
```suggestion
  let fileStr := "\n".intercalate lines.toList
```

Reply:
lol, I don' t know why I left that TODO instead of doing it :thinking: